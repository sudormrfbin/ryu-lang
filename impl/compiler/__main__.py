from compiler.repl import repl

repl()

# with open("input.lang", "r") as f:
#     inp = f.read()

# output = compile(inp)
# print(output)
