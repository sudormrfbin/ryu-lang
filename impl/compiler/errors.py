from typing import ClassVar
import dataclasses
from dataclasses import dataclass
from typing_extensions import override

from lark.tree import Meta
from lark.lexer import Token

from compiler import langtypes

LineCol = tuple[int, int]


@dataclass
class Span:
    start_line: int
    end_line: int
    start_column: int
    end_column: int
    start_pos: int
    end_pos: int

    @classmethod
    def from_meta(cls, meta: Meta) -> "Span":
        return cls(
            start_line=meta.line,
            end_line=meta.end_line,
            start_column=meta.column,
            end_column=meta.end_column,
            start_pos=meta.start_pos,
            end_pos=meta.end_pos,
        )

    @classmethod
    def from_token(cls, token: Token) -> "Span":
        assert token.line is not None
        assert token.end_line is not None
        assert token.column is not None
        assert token.end_column is not None
        assert token.start_pos is not None
        assert token.end_pos is not None

        return cls(
            start_line=token.line,
            end_line=token.end_line,
            start_column=token.column,
            end_column=token.end_column,
            start_pos=token.start_pos,
            end_pos=token.end_pos,
        )

    def coord(self) -> tuple[LineCol, LineCol]:
        start = (self.start_line, self.start_column)
        end = (self.end_line, self.end_column)
        return (start, end)


class InternalCompilerError(Exception):
    pass


@dataclass
class CompilerError(Exception):
    code: ClassVar[int]

    message: str
    span: Span

    def __post_init__(self):
        attributes = dataclasses.astuple(self)
        # Custom exceptions must always call super(), we use the __post_init__
        # mechanism provided by dataclasses to call the super constructor after
        # this class's __init__ is called.
        super().__init__(*attributes)

    @override
    def __str__(self) -> str:
        return f"{type(self).__name__}: {self.message}"


@dataclass
class OperatorSpan:
    name: str
    span: Span


@dataclass
class OperandSpan:
    type_: langtypes.Type
    span: Span


@dataclass
class InvalidOperationError(CompilerError):
    code = 1

    operator: OperatorSpan
    operands: list[OperandSpan]


@dataclass
class UnknownVariable(CompilerError):
    """
    Raised when a variable is used in an expression without declaring it prior.

    ## Example
    ```
    print(var + 5)
    ```
    """

    code = 2

    variable: str


@dataclass
class UndeclaredVariable(CompilerError):
    """
    Raised when the lvalue (the variable part) of an assignment has not
    been declared prior to assignment.

    ## Example
    ```
    var = 4
    ```

    ## Fix
    ```
    let var = 4
    ```
    """

    code = 3

    variable: str


@dataclass
class UnexpectedType(CompilerError):
    """
    Raised when an expression deviates from it's expected type.

    ## Example
    ```
    if 1 {}
    ```

    ## Fix
    ```
    if true {}
    ```
    """

    code = 4

    expected_type: langtypes.Type
    actual_type: langtypes.Type


@dataclass
class TypeMismatch(CompilerError):
    """
    Raised when an expression is expected to be of a particular type
    because of a declaration in another point in the source code.

    ## Example
    ```
    match true {
        case 1 {}
        case 2 {}
    }
    ```

    ## Fix
    ```
    match 1 {
        case 1 {}
        case 2 {}
    }
    ```
    """

    code = 5

    actual_type: langtypes.Type
    expected_type: langtypes.Type
    expected_type_span: Span


@dataclass
class DuplicatedCase(CompilerError):
    """
    Raised when a match statement has duplicated case arms.

    ## Example
    ```
    match true {
        case true {}
        case false {}
    }
    ```

    ## Fix
    ```
    match true {
        case true {}
        case false {}
    }
    ```
    """

    code = 6

    previous_case_span: Span
