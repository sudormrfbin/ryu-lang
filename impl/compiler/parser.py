from typing import Any

from compiler import ast
from compiler.lalr import Token, Tree, Lark_StandAlone, v_args, Transformer, DATA

# https://github.com/lark-parser/lark/issues/565
DATA["options"]["maybe_placeholders"] = True
_parser = Lark_StandAlone(propagate_positions=True)


def parse(source: str) -> Tree[Token]:
    return _parser.parse(source)


def listify(cls):
    @v_args(meta=True, inline=False)
    def _(self, meta, lst):
        return cls(meta, lst)

    return _


@v_args(meta=True, inline=True)
class LarkTreeToAstTransformer(Transformer[Token, Any]):
    def TRUE(self, _) -> bool:
        return True

    def FALSE(self, _) -> bool:
        return False

    def INT(self, n: str) -> int:
        return int(n)

    def STRING(self, s: str) -> str:
        return s[1:-1]  # remove quotes

    statement_list = listify(ast.StatementList)
    statement_block = listify(ast.StatementBlock)
    variable_declaration = ast.VariableDeclaration
    assignment = ast.Assignment
    index_assignment = ast.IndexAssignment
    print_stmt = ast.PrintStmt
    if_chain = ast.IfChain
    if_stmt = ast.IfStmt
    else_if_ladder = listify(ast.ElseIfLadder)
    else_if_stmt = ast.ElseIfStmt
    match_stmt = ast.MatchStmt
    case_ladder = listify(ast.CaseLadder)
    case_stmt = ast.CaseStmt
    enum_pattern = ast.EnumPattern
    enum_pattern_tuple = ast.EnumPatternTuple
    wildcard_pattern = ast.WildcardPattern
    while_stmt = ast.WhileStmt
    for_stmt = ast.ForStmt
    for_stmt_int = ast.ForStmtInt
    struct_stmt = ast.StructStmt
    struct_members = listify(ast.StructMembers)
    struct_member = ast.StructMember
    struct_access = ast.StructAccess
    struct_assignment = ast.StructAssignment
    enum_stmt = ast.EnumStmt
    enum_members = listify(ast.EnumMembers)
    enum_member_bare = ast.EnumMemberBare
    enum_member_tuple = ast.EnumMemberTuple
    function_definition = ast.FunctionDefinition
    function_params = listify(ast.FunctionParams)
    function_param = ast.FunctionParam
    type_annotation = ast.TypeAnnotation
    return_stmt = ast.ReturnStmt
    equality = ast.Equality
    logical = ast.Logical
    comparison = ast.Comparison
    term = ast.Term
    factor = ast.Factor
    unary_op = ast.UnaryOp
    indexing = ast.Indexing
    function_call = ast.FunctionCall
    function_args = listify(ast.FunctionArgs)
    struct_init_members = listify(ast.StructInitMembers)
    struct_init_member = ast.StructInitMember
    bool_literal = ast.BoolLiteral
    int_literal = ast.IntLiteral
    string_literal = ast.StringLiteral
    enum_literal_simple = ast.EnumLiteralSimple
    enum_literal_tuple = ast.EnumLiteralTuple
    variable = ast.Variable
    array_literal = ast.ArrayLiteral
    array_elements = listify(ast.ArrayElements)
    array_element = ast.ArrayElement
    array_pattern = listify(ast.ArrayPattern)
    array_pattern_element = ast.ArrayPatternElement


_transformer = LarkTreeToAstTransformer()


def parse_tree_to_ast(tree: Tree[Token]) -> ast.Ast:
    return _transformer.transform(tree)
