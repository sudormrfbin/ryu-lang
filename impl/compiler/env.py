from typing import Any

from .errors import InternalCompilerError


class RuntimeEnvironment:
    values: dict[str, Any]

    def __init__(self):
        self.values = {}

    def define(self, name: str, value: Any):
        # shadowing is allowed
        self.values[name] = value

    def get(self, name: str) -> Any:
        try:
            return self.values[name]
        except KeyError:
            raise InternalCompilerError(f"Variable {name} not defined at runtime")
