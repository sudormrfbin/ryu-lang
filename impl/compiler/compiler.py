from typing import Any

from compiler.env import RuntimeEnvironment, TypeEnvironment

from compiler import builtins, errors, langvalues
from compiler.parser import parse, parse_tree_to_ast

BUILTIN_FUNCTIONS: list[type[langvalues.BuiltinFunction]] = [
    builtins.SumFunction,
    builtins.ArrayLengthFunction,
    builtins.StringLengthFunction,
    builtins.ArrayAppend,
]


def get_default_environs() -> tuple[TypeEnvironment, RuntimeEnvironment]:
    type_env = TypeEnvironment()
    runtime_env = RuntimeEnvironment()

    for fn in BUILTIN_FUNCTIONS:
        type_env.define(fn.TYPE.function_name, fn.TYPE)
        runtime_env.define(fn.TYPE.function_name, fn())

    return (type_env, runtime_env)


def run(source: str, type_env: TypeEnvironment, runtime_env: RuntimeEnvironment) -> Any:
    try:
        return _run(source, type_env, runtime_env)
    except errors.CompilerError as err:
        err.report(source)


def _run(
    source: str, type_env: TypeEnvironment, runtime_env: RuntimeEnvironment
) -> Any:
    tree = parse(source)
    ast = parse_tree_to_ast(tree)

    ast.typecheck(type_env)

    return ast.eval(runtime_env)
