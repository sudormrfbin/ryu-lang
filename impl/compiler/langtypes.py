from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional, Sequence, TypeAlias
from typing_extensions import override


if TYPE_CHECKING:
    from compiler.env import TypeEnvironment
    from compiler.errors import Span


class Type:
    @property
    def name(self) -> str:
        return type(self).__name__

    @classmethod
    def from_str(cls, ident: str, env: TypeEnvironment) -> Optional["Type"]:
        if ident in PRIMITIVE_TYPES:
            return PRIMITIVE_TYPES[ident]

        ty = env.get(ident)
        if isinstance(ty, UserDefinedType):
            return ty

        return None

    @classmethod
    def from_str_with_generics(
        cls,
        ident: str,
        generics: Type,
        env: TypeEnvironment,
    ) -> Optional["Type"]:
        if ident == "array":
            return Array(generics)

        return None


class Placeholder(Type):
    pass


class PrimitiveType(Type):
    pass


class UserDefinedType(Type):
    pass


class Bool(PrimitiveType):
    pass


class Int(PrimitiveType):
    pass


class String(PrimitiveType):
    pass


class Block(Type):
    pass


@dataclass
class ReturnBlock(Block):
    return_type: Type
    return_stmt_span: Span


@dataclass
class Struct(UserDefinedType):
    struct_name: str
    members: Members

    @dataclass
    class Members(PrimitiveType):
        types: dict[str, Type]

        def __len__(self) -> int:
            return len(self.types)


@dataclass
class Enum(UserDefinedType):
    enum_name: str
    members: list[EnumVariantSimple | EnumVariantTuple]
    span: Span

    @property
    @override
    def name(self) -> str:
        return self.enum_name

    def variant_from_str(
        self, name: str
    ) -> EnumVariantSimple | EnumVariantTuple | None:
        for mem in self.members:
            if mem.name == name:
                return mem
        return None


@dataclass
class EnumVariantSimple:
    name: str


@dataclass
class EnumVariantTuple:
    name: str
    inner: Type


EnumVariants: TypeAlias = EnumVariantSimple | EnumVariantTuple


@dataclass
class Array(PrimitiveType):
    ty: Type

    @property
    @override
    def name(self) -> str:
        return f"Array<{self.ty.name}>"


@dataclass
class UntypedArray(PrimitiveType):
    pass


@dataclass
class Function(PrimitiveType):
    function_name: str
    arguments: Params
    return_type: Type


@dataclass
class Params(PrimitiveType):
    types: list[Type]

    def __len__(self) -> int:
        return len(self.types)


BOOL = Bool()
INT = Int()
STRING = String()
BLOCK = Block()
PLACEHOLDER = Placeholder()


PRIMITIVE_TYPES: dict[str, Type] = {
    "bool": BOOL,
    "int": INT,
    "string": STRING,
}


def resolve_blocks_type(types: Sequence[Type]) -> Block:
    resolved = BLOCK
    for ty in types:
        match (resolved, ty):
            case (Block(), ReturnBlock()):
                resolved = ty
            case (ReturnBlock(), ReturnBlock()):
                if resolved.return_type != ty.return_type:
                    raise  # TODO different return types
            case _:
                pass

    return resolved
