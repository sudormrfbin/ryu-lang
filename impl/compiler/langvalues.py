from abc import abstractmethod
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any
from typing_extensions import override
from compiler import runtime

from compiler.env import RuntimeEnvironment

if TYPE_CHECKING:
    from compiler.ast import StatementBlock


@dataclass
class EnumValue:
    ty: str
    variant: str

    @override
    def __str__(self) -> str:
        return f"{self.ty}::{self.variant}"

    @override
    def __hash__(self) -> int:
        return hash(repr(self))


class Function:
    @abstractmethod
    def call(self, args: list[Any], env: "RuntimeEnvironment") -> Any:
        pass


@dataclass
class RyuFunction(Function):
    param_names: list[str]
    body: "StatementBlock"

    @override
    def call(self, args: list[Any], env: "RuntimeEnvironment") -> Any:
        child_env = RuntimeEnvironment(enclosing=env)

        for name, arg in zip(self.param_names, args):
            child_env.define(name, arg)

        try:
            self.body.eval(child_env)
        except runtime.FunctionReturn as ret:
            return ret.return_value
