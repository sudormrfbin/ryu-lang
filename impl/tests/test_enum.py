from typing import Any
from compiler import langtypes
from compiler.compiler import get_default_environs
from compiler.env import RuntimeEnvironment, TypeEnvironment
from compiler.errors import Span
from compiler.langtypes import BOOL, INT, Enum
from compiler.langvalues import EnumValue
from compiler.parser import parse, parse_tree_to_ast
from tests.utils import docstring_source_with_snapshot


@docstring_source_with_snapshot
def test_enum_def(source: str, snapshot: Any):
    """
    enum Langs {
        Malayalam
        English
        Japanese
    }
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot

    type_env = TypeEnvironment()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == snapshot

    assert type_env.get_type("Langs") == Enum(
        enum_name="Langs",
        members=[
            Enum.Simple("Malayalam"),
            Enum.Simple("English"),
            Enum.Simple("Japanese"),
        ],
        span=Span(
            start_line=1,
            end_line=1,
            start_column=6,
            end_column=11,
            start_pos=5,
            end_pos=10,
        ),
    )


@docstring_source_with_snapshot
def test_enum_assignment(source: str, snapshot: Any):
    """
    enum Langs {
        Malayalam
        English
        Japanese
    }
    let lang = Langs::English
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot

    type_env = TypeEnvironment()
    ast.typecheck(type_env)
    print(ast.to_type_dict())
    assert ast.to_type_dict() == snapshot

    lang_type = Enum(
        enum_name="Langs",
        members=[
            Enum.Simple("Malayalam"),
            Enum.Simple("English"),
            Enum.Simple("Japanese"),
        ],
        span=Span(
            start_line=1,
            end_line=1,
            start_column=6,
            end_column=11,
            start_pos=5,
            end_pos=10,
        ),
    )
    assert type_env.get_type("Langs") == lang_type
    assert type_env.get_var_type("lang") == lang_type

    env = RuntimeEnvironment()
    ast.eval(env)
    assert env.get("lang") == EnumValue(ty="Langs", variant="English")


@docstring_source_with_snapshot
def test_enum_pattern_match(source: str, snapshot: Any):
    """
    enum Langs {
        Malayalam
        English
        Japanese
    }
    let lang = Langs::English

    let langcode = ""
    match lang {
        case Langs::English { langcode = "eng" }
        case Langs::Malayalam { langcode = "ml" }
        case Langs::Japanese { langcode = "jp" }
    }
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot

    type_env = TypeEnvironment()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == snapshot

    lang_type = Enum(
        enum_name="Langs",
        members=[
            Enum.Simple("Malayalam"),
            Enum.Simple("English"),
            Enum.Simple("Japanese"),
        ],
        span=Span(
            start_line=1,
            end_line=1,
            start_column=6,
            end_column=11,
            start_pos=5,
            end_pos=10,
        ),
    )
    assert type_env.get_type("Langs") == lang_type
    assert type_env.get_var_type("lang") == lang_type
    assert type_env.get_var_type("langcode") == langtypes.STRING

    env = RuntimeEnvironment()
    ast.eval(env)
    assert env.get("lang") == EnumValue(ty="Langs", variant="English")
    assert env.get("langcode") == "eng"


@docstring_source_with_snapshot
def test_tuple_enum_def(source: str, snapshot: Any):
    """
    enum MaybeInt {
        Some(int)
        None
    }
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot

    type_env, _ = get_default_environs()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == snapshot

    assert type_env.get_type("MaybeInt") == Enum(
        enum_name="MaybeInt",
        members=[Enum.Tuple("Some", INT), Enum.Simple("None")],
        span=Span(
            start_line=1,
            end_line=1,
            start_column=6,
            end_column=14,
            start_pos=5,
            end_pos=13,
        ),
    )


@docstring_source_with_snapshot
def test_tuple_enum_assignment(source: str, snapshot: Any):
    """
    enum MaybeInt {
        Some(int)
        None
    }

    let x = MaybeInt::Some(8)
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot

    type_env, env = get_default_environs()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == snapshot

    enum_type = Enum(
        enum_name="MaybeInt",
        members=[Enum.Tuple("Some", INT), Enum.Simple("None")],
        span=Span(
            start_line=1,
            end_line=1,
            start_column=6,
            end_column=14,
            start_pos=5,
            end_pos=13,
        ),
    )

    assert type_env.get_type("MaybeInt") == enum_type
    assert type_env.get_var_type("x") == enum_type

    ast.eval(env)


@docstring_source_with_snapshot
def test_tuple_enum_match(source: str, snapshot: Any):
    """
    enum MaybeBool {
        Some(bool)
        None
    }

    fn eval(v: MaybeBool) -> int {
        match v {
            case MaybeBool::Some(true) { return 0 }
            case MaybeBool::Some(false) { return 1 }
            case MaybeBool::None { return 2 }
        }
    }

    let zero = eval(MaybeBool::Some(true))
    let one = eval(MaybeBool::Some(false))
    let two = eval(MaybeBool::None)
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot

    type_env, env = get_default_environs()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == snapshot

    enum_type = Enum(
        enum_name="MaybeBool",
        members=[Enum.Tuple("Some", BOOL), Enum.Simple("None")],
        span=Span(
            start_line=1,
            end_line=1,
            start_column=6,
            end_column=15,
            start_pos=5,
            end_pos=14,
        ),
    )

    assert type_env.get_type("MaybeBool") == enum_type

    ast.eval(env)

    assert env.get("zero") == 0
    assert env.get("one") == 1
    assert env.get("two") == 2


# @docstring_source_with_snapshot
# def test_enum_with_generics(source: str, snapshot: Any):
#     """
#     enum Option<T> {
#         Some(T)
#         None
#     }

#     enum Result<T, E> {
#         Ok(T),
#         Err(E)
#     }
#     """
#     ast = parse_tree_to_ast(parse(source))
#     assert ast.to_dict() == snapshot

#     type_env = TypeEnvironment()
#     ast.typecheck(type_env)
#     assert ast.to_type_dict() == snapshot

#     assert type_env.get("Option") == Enum(
#         enum_name="Option",
#         members=["Some", "None"],
#         span=Span(
#             start_line=1,
#             end_line=1,
#             start_column=6,
#             end_column=11,
#             start_pos=5,
#             end_pos=10,
#         ),
#     )
