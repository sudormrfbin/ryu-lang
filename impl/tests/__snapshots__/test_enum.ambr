# serializer version: 1
# name: test_enum_assignment
  dict({
    <class 'compiler.ast.StatementList'>: dict({
      'stmts': list([
        dict({
          <class 'compiler.ast.EnumStmt'>: dict({
            'members': dict({
              <class 'compiler.ast.EnumMembers'>: dict({
                'members': list([
                  dict({
                    <class 'compiler.ast.EnumMemberBare'>: dict({
                      'name': Token('IDENTIFIER', 'Malayalam'),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.EnumMemberBare'>: dict({
                      'name': Token('IDENTIFIER', 'English'),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.EnumMemberBare'>: dict({
                      'name': Token('IDENTIFIER', 'Japanese'),
                    }),
                  }),
                ]),
              }),
            }),
            'name': Token('IDENTIFIER', 'Langs'),
          }),
        }),
        dict({
          <class 'compiler.ast.VariableDeclaration'>: dict({
            'ident': Token('IDENTIFIER', 'lang'),
            'rvalue': dict({
              <class 'compiler.ast.EnumLiteralSimple'>: dict({
                'enum_type': Token('IDENTIFIER', 'Langs'),
                'variant': Token('IDENTIFIER', 'English'),
              }),
            }),
          }),
        }),
      ]),
    }),
  })
# ---
# name: test_enum_assignment.1
  dict({
    'stmts': list([
      dict({
        'members': dict({
          'members': list([
            dict({
              <class 'compiler.ast.EnumMemberBare'>: <class 'compiler.langtypes.Type'>,
            }),
            dict({
              <class 'compiler.ast.EnumMemberBare'>: <class 'compiler.langtypes.Type'>,
            }),
            dict({
              <class 'compiler.ast.EnumMemberBare'>: <class 'compiler.langtypes.Type'>,
            }),
          ]),
          <class 'compiler.ast.EnumMembers'>: <class 'compiler.langtypes.Block'>,
        }),
        <class 'compiler.ast.EnumStmt'>: <class 'compiler.langtypes.Enum'>,
      }),
      dict({
        'rvalue': dict({
          <class 'compiler.ast.EnumLiteralSimple'>: <class 'compiler.langtypes.Enum'>,
        }),
        <class 'compiler.ast.VariableDeclaration'>: <class 'compiler.langtypes.Enum'>,
      }),
    ]),
    <class 'compiler.ast.StatementList'>: <class 'compiler.langtypes.Block'>,
  })
# ---
# name: test_enum_def
  dict({
    <class 'compiler.ast.EnumStmt'>: dict({
      'members': dict({
        <class 'compiler.ast.EnumMembers'>: dict({
          'members': list([
            dict({
              <class 'compiler.ast.EnumMemberBare'>: dict({
                'name': Token('IDENTIFIER', 'Malayalam'),
              }),
            }),
            dict({
              <class 'compiler.ast.EnumMemberBare'>: dict({
                'name': Token('IDENTIFIER', 'English'),
              }),
            }),
            dict({
              <class 'compiler.ast.EnumMemberBare'>: dict({
                'name': Token('IDENTIFIER', 'Japanese'),
              }),
            }),
          ]),
        }),
      }),
      'name': Token('IDENTIFIER', 'Langs'),
    }),
  })
# ---
# name: test_enum_def.1
  dict({
    'members': dict({
      'members': list([
        dict({
          <class 'compiler.ast.EnumMemberBare'>: <class 'compiler.langtypes.Type'>,
        }),
        dict({
          <class 'compiler.ast.EnumMemberBare'>: <class 'compiler.langtypes.Type'>,
        }),
        dict({
          <class 'compiler.ast.EnumMemberBare'>: <class 'compiler.langtypes.Type'>,
        }),
      ]),
      <class 'compiler.ast.EnumMembers'>: <class 'compiler.langtypes.Block'>,
    }),
    <class 'compiler.ast.EnumStmt'>: <class 'compiler.langtypes.Enum'>,
  })
# ---
# name: test_enum_pattern_match
  dict({
    <class 'compiler.ast.StatementList'>: dict({
      'stmts': list([
        dict({
          <class 'compiler.ast.EnumStmt'>: dict({
            'members': dict({
              <class 'compiler.ast.EnumMembers'>: dict({
                'members': list([
                  dict({
                    <class 'compiler.ast.EnumMemberBare'>: dict({
                      'name': Token('IDENTIFIER', 'Malayalam'),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.EnumMemberBare'>: dict({
                      'name': Token('IDENTIFIER', 'English'),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.EnumMemberBare'>: dict({
                      'name': Token('IDENTIFIER', 'Japanese'),
                    }),
                  }),
                ]),
              }),
            }),
            'name': Token('IDENTIFIER', 'Langs'),
          }),
        }),
        dict({
          <class 'compiler.ast.VariableDeclaration'>: dict({
            'ident': Token('IDENTIFIER', 'lang'),
            'rvalue': dict({
              <class 'compiler.ast.EnumLiteralSimple'>: dict({
                'enum_type': Token('IDENTIFIER', 'Langs'),
                'variant': Token('IDENTIFIER', 'English'),
              }),
            }),
          }),
        }),
        dict({
          <class 'compiler.ast.VariableDeclaration'>: dict({
            'ident': Token('IDENTIFIER', 'langcode'),
            'rvalue': dict({
              <class 'compiler.ast.StringLiteral'>: dict({
                'value': '',
              }),
            }),
          }),
        }),
        dict({
          <class 'compiler.ast.MatchStmt'>: dict({
            'cases': dict({
              <class 'compiler.ast.CaseLadder'>: dict({
                'cases': list([
                  dict({
                    <class 'compiler.ast.CaseStmt'>: dict({
                      'block': dict({
                        <class 'compiler.ast.StatementBlock'>: dict({
                          'stmts': list([
                            dict({
                              <class 'compiler.ast.Assignment'>: dict({
                                'lvalue': Token('IDENTIFIER', 'langcode'),
                                'rvalue': dict({
                                  <class 'compiler.ast.StringLiteral'>: dict({
                                    'value': 'eng',
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                      'pattern': dict({
                        <class 'compiler.ast.EnumPattern'>: dict({
                          'enum_type': Token('IDENTIFIER', 'Langs'),
                          'variant': Token('IDENTIFIER', 'English'),
                        }),
                      }),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.CaseStmt'>: dict({
                      'block': dict({
                        <class 'compiler.ast.StatementBlock'>: dict({
                          'stmts': list([
                            dict({
                              <class 'compiler.ast.Assignment'>: dict({
                                'lvalue': Token('IDENTIFIER', 'langcode'),
                                'rvalue': dict({
                                  <class 'compiler.ast.StringLiteral'>: dict({
                                    'value': 'ml',
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                      'pattern': dict({
                        <class 'compiler.ast.EnumPattern'>: dict({
                          'enum_type': Token('IDENTIFIER', 'Langs'),
                          'variant': Token('IDENTIFIER', 'Malayalam'),
                        }),
                      }),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.CaseStmt'>: dict({
                      'block': dict({
                        <class 'compiler.ast.StatementBlock'>: dict({
                          'stmts': list([
                            dict({
                              <class 'compiler.ast.Assignment'>: dict({
                                'lvalue': Token('IDENTIFIER', 'langcode'),
                                'rvalue': dict({
                                  <class 'compiler.ast.StringLiteral'>: dict({
                                    'value': 'jp',
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                      'pattern': dict({
                        <class 'compiler.ast.EnumPattern'>: dict({
                          'enum_type': Token('IDENTIFIER', 'Langs'),
                          'variant': Token('IDENTIFIER', 'Japanese'),
                        }),
                      }),
                    }),
                  }),
                ]),
              }),
            }),
            'expr': dict({
              <class 'compiler.ast.Variable'>: dict({
                'value': Token('IDENTIFIER', 'lang'),
              }),
            }),
          }),
        }),
      ]),
    }),
  })
# ---
# name: test_enum_pattern_match.1
  dict({
    'stmts': list([
      dict({
        'members': dict({
          'members': list([
            dict({
              <class 'compiler.ast.EnumMemberBare'>: <class 'compiler.langtypes.Type'>,
            }),
            dict({
              <class 'compiler.ast.EnumMemberBare'>: <class 'compiler.langtypes.Type'>,
            }),
            dict({
              <class 'compiler.ast.EnumMemberBare'>: <class 'compiler.langtypes.Type'>,
            }),
          ]),
          <class 'compiler.ast.EnumMembers'>: <class 'compiler.langtypes.Block'>,
        }),
        <class 'compiler.ast.EnumStmt'>: <class 'compiler.langtypes.Enum'>,
      }),
      dict({
        'rvalue': dict({
          <class 'compiler.ast.EnumLiteralSimple'>: <class 'compiler.langtypes.Enum'>,
        }),
        <class 'compiler.ast.VariableDeclaration'>: <class 'compiler.langtypes.Enum'>,
      }),
      dict({
        'rvalue': dict({
          <class 'compiler.ast.StringLiteral'>: <class 'compiler.langtypes.String'>,
        }),
        <class 'compiler.ast.VariableDeclaration'>: <class 'compiler.langtypes.String'>,
      }),
      dict({
        'cases': dict({
          'cases': list([
            dict({
              'block': dict({
                'stmts': list([
                  dict({
                    'rvalue': dict({
                      <class 'compiler.ast.StringLiteral'>: <class 'compiler.langtypes.String'>,
                    }),
                    <class 'compiler.ast.Assignment'>: <class 'compiler.langtypes.String'>,
                  }),
                ]),
                <class 'compiler.ast.StatementBlock'>: <class 'compiler.langtypes.Block'>,
              }),
              'pattern': dict({
                <class 'compiler.ast.EnumPattern'>: <class 'compiler.langtypes.Enum'>,
              }),
              <class 'compiler.ast.CaseStmt'>: <class 'compiler.langtypes.Enum'>,
            }),
            dict({
              'block': dict({
                'stmts': list([
                  dict({
                    'rvalue': dict({
                      <class 'compiler.ast.StringLiteral'>: <class 'compiler.langtypes.String'>,
                    }),
                    <class 'compiler.ast.Assignment'>: <class 'compiler.langtypes.String'>,
                  }),
                ]),
                <class 'compiler.ast.StatementBlock'>: <class 'compiler.langtypes.Block'>,
              }),
              'pattern': dict({
                <class 'compiler.ast.EnumPattern'>: <class 'compiler.langtypes.Enum'>,
              }),
              <class 'compiler.ast.CaseStmt'>: <class 'compiler.langtypes.Enum'>,
            }),
            dict({
              'block': dict({
                'stmts': list([
                  dict({
                    'rvalue': dict({
                      <class 'compiler.ast.StringLiteral'>: <class 'compiler.langtypes.String'>,
                    }),
                    <class 'compiler.ast.Assignment'>: <class 'compiler.langtypes.String'>,
                  }),
                ]),
                <class 'compiler.ast.StatementBlock'>: <class 'compiler.langtypes.Block'>,
              }),
              'pattern': dict({
                <class 'compiler.ast.EnumPattern'>: <class 'compiler.langtypes.Enum'>,
              }),
              <class 'compiler.ast.CaseStmt'>: <class 'compiler.langtypes.Enum'>,
            }),
          ]),
          <class 'compiler.ast.CaseLadder'>: <class 'compiler.langtypes.Block'>,
        }),
        'expr': dict({
          <class 'compiler.ast.Variable'>: <class 'compiler.langtypes.Enum'>,
        }),
        <class 'compiler.ast.MatchStmt'>: <class 'compiler.langtypes.Block'>,
      }),
    ]),
    <class 'compiler.ast.StatementList'>: <class 'compiler.langtypes.Block'>,
  })
# ---
# name: test_tuple_enum_def
  dict({
    <class 'compiler.ast.EnumStmt'>: dict({
      'members': dict({
        <class 'compiler.ast.EnumMembers'>: dict({
          'members': list([
            dict({
              <class 'compiler.ast.EnumMemberTuple'>: dict({
                'name': Token('IDENTIFIER', 'Some'),
                'tuple_members': dict({
                  <class 'compiler.ast.TypeAnnotation'>: dict({
                    'ty': Token('IDENTIFIER', 'int'),
                  }),
                }),
              }),
            }),
            dict({
              <class 'compiler.ast.EnumMemberBare'>: dict({
                'name': Token('IDENTIFIER', 'None'),
              }),
            }),
          ]),
        }),
      }),
      'name': Token('IDENTIFIER', 'MaybeInt'),
    }),
  })
# ---
# name: test_tuple_enum_def.1
  dict({
    'members': dict({
      'members': list([
        dict({
          'tuple_members': dict({
            <class 'compiler.ast.TypeAnnotation'>: <class 'compiler.langtypes.Int'>,
          }),
          <class 'compiler.ast.EnumMemberTuple'>: <class 'compiler.langtypes.Type'>,
        }),
        dict({
          <class 'compiler.ast.EnumMemberBare'>: <class 'compiler.langtypes.Type'>,
        }),
      ]),
      <class 'compiler.ast.EnumMembers'>: <class 'compiler.langtypes.Block'>,
    }),
    <class 'compiler.ast.EnumStmt'>: <class 'compiler.langtypes.Enum'>,
  })
# ---
